transmogrifier/models.py

```diff
@@ -1,25 +1,31 @@
-from typing import Optional
+from collections.abc import Callable
+from typing import Any
 
+import attrs
 from attrs import asdict, define, field, validators
 from attrs.validators import instance_of, optional
 
 
-def check_range(instance, attribute, value):
+def check_range(
+    instance: "Date", attribute: "attrs.Attribute", value: "DateRange"  # noqa: ARG001
+) -> None:
     if value is None:
         return
     if value.gt and value.gte:
-        raise ValueError(
+        message = (
             f"{attribute.name} may have a 'gt' or 'gte' value, but not both; "
             f"received {value}"
         )
+        raise ValueError(message)
     if value.lt and value.lte:
-        raise ValueError(
+        message = (
             f"{attribute.name} may have a 'lt' or 'lte' value, but not both; "
             f"received {value}"
         )
+        raise ValueError(message)
 
 
-def list_of(item_type):
+def list_of(item_type: Any) -> Callable:  # noqa: ANN401
```

Comment:
I think `Any` is a good call here. Sure, we could scan the codebase and create a fairly large union of possible types... but we don't gain much from that.